java.util.LinkedList spec draft

alexeev509 opened the pull request:
Was added first version of the LinkedList LibSL realization. Some parts of this will be changed.

alexeev509 committed:
Create LinkedList.lsl

alexeev509 committed:
Was added LinkedList automaton

Was added LinkedList automaton with definite annotations: @Public, @Extends, @Implements, @WrapperMeta

alexeev509 committed:
Add automaton variables

Were added automaton variables

alexeev509 committed:
Were added constructors

Attention: we have problem in the second constructor, because we don't know how to avoid cycle in the method addAll for another collection

alexeev509 committed:
Were added first 10 methods implementation

alexeev509 committed:
7 methods were added

alexeev509 committed:
Were added many methods

alexeev509 committed:
LIST_REMOVE Update

alexeev509 committed:
Was created first realization ListItr automaton

Not implemented forEachRemaining method, because we don't know what to do with Lamdas

alexeev509 committed:
LIST_REMOVE

Was added new action: LIST_REMOVE

alexeev509 committed:
States of the List

Were added states of the LinkedList

alexeev509 committed:
Sub

Were changed some fun's to sub's.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+automaton LinkedList: int(
+   	var storage: list<Object>,
+	@Transient var size: int = 0,
+	@Protected @Transient var modCount: int = 0,
```
The `protected` annotation is not needed here. It can be used later when we add automata inheritance or something similar.

alexeev509 replied:
Has been fixed

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+    matchInterfaces=true,
+)
+automaton LinkedList: int(
+   	var storage: list<Object>,
```
Please use consisting spacing: 4 Space symbols per 1 Tab symbol.

alexeev509 replied:
Has been fixed

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+   	var storage: list<Object>,
+	@Transient var size: int = 0,
+	@Protected @Transient var modCount: int = 0,
+    	var serialVersionUID:long = 876323262645176354
```
Add `const` or `final` annotation modifier if necessary.

alexeev509 replied:
Has been fixed. Was added @Final annotation to automaton fields.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+
+	//constructors
+
+	constructor LinkedList (): void {
```
Constructors, as it have been discussed lately, do not have return value, so please remove return type specifier.

alexeev509 replied:
Has been fixed

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+
+
+	fun getLast() : Object {
+		if (size==0)
```
Please add spacing around operators.

alexeev509 replied:
Has been fixed

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+		action LIST_REMOVE(storage, index);
+		//Problem
+		//We need add decrement and increment in the LibSL
+		size--;
```
We decided to postpone uses of pre- and post- increments, so please do not use such functionality in the near future.

alexeev509 replied:
Has been fixed; All decremenets (and increments) were replaced in such way: "size = size - 1";

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	}
+
+
+	@Private
```
Subprograms should (I hope) generate `private` by default (I hope), so annotation is not necessary here.

alexeev509 replied:
Has been fixed

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	fun remove (o: Object): boolean {
+		var index = indexOf(o);
+		result = action LIST_REMOVE(storage, index);
+		*Можно ли писать result НЕ в самом конце ? *
```
Please use comment syntax for actual comments and questions.
And, yes, variable `result` can be accessed (and be overwritten) anywhere inside of the method/function body. It's **not** a `return` statement.

alexeev509 replied:
Thank you @dog-m , I forgot to delete/edit comments in some places. Has been fixed

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	}
+	
+	
+	fun addAll (index:int, c:Collection): boolean {
```
Consider adding spacing *between* `:` symbol and type specifiers.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	
+	fun addAll (index:int, c:Collection): boolean {
+		//problem:
+		//we don't know how to avoid cycle i this method
```
Please use `action NOT_IMPLEMENTED();` to indicate functionality that have not been implemented for some reason.

alexeev509 replied:
Has been fixed

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	
+	
+	fun removeLastOccurrence(o: Object): boolean {
+		//I need think about this
```
Please use `action NOT_IMPLEMENTED();` to indicate functionality that have not been implemented for some reason.

alexeev509 replied:
Has been fixed

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	fun spliterator(): Spliterator {
+		result = new LLSpliterator(state=Initialized,
+		//This is right ? "parent=self"
+		parent=self,
```
Please use additional spacing in case of complex object creation when multiple lines used for specifying variable states.

alexeev509 replied:
Has been fixed.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	var nextIndex:int,
+	var next: Object,
+	var lastReturned: Object,
+	var expectedModCount = parent.modCount)
```
I don't think such functionality is supported or even mentioned during last discussions.

alexeev509 replied:
Has been fixed.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+{
+	initstate Initialized;
+	
+	shift Initialized -> self by [
```
Please use consisting spacing.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	{
+		//We need to insert before next
+		var index = self.parent.indexof(next) - 1;
+		self.parent.linkAny(, e);
```
First argument is missing.

alexeev509 replied:
Has been fixed

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	lastReturned = null;
+	if(next == null)
+	{
+		self.parent.linkAny(self.parent-size - 1, e);
```
Incorrect field access separator `-` (dash). Use `.` (dot) instead.

alexeev509 replied:
Has been fixed

dog-m commented on spec/list-actions.lsl:
```diff
@@ -35,4 +35,8 @@ define action LIST_RESIZE(
+    
+define action LIST_REMOVE(
+        aList: list<any>,
+        index: int
```
I suggest adding `count` parameter to extend functionality of this action on cases when a continuous group of multiple items being removed at the same time.

alexeev509 replied:
Was added count param.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+    
+    constructor ListItr(index: int): void 
+    {
+    	if(index == self.parent.size)
```
But what will happen if `index` is greater than `size`? (a hint)

alexeev509 replied:
Has been fixed.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+)
+automaton ListItr: int(
+	var nextIndex:int,
+	var next: Object,
```
Variables with references to arbitrary objects are not allowed, as it have been discussed lately.

alexeev509 replied:
Has been fixed. Field "next" was replaced by two flags "nextWasCalled" and "prevWasCalled".

alexeev509 committed:
DescendingIterator

Was added new automaton: DescendingIterator.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+    
+    //constructors
+    
+    constructor ListItr(index: int): void 
```
The constructor is a public function, so it should (I guess) be also mentioned in state transitions.

alexeev509 replied:
This constructor was added to state transitions.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -0,0 +1,666 @@
+	//problem:
+	// we need add constraint for collection type: Collection<? extends E> 
+	fun addAll (c: Collection): boolean {
+		result = addAll(size, c);
```
Calls to public `fun`ctions are not allowed inside of other `fun`ctions. Consider adding a subroutine of common code that is being used.

alexeev509 replied:
@dog-m It was added. But can we invoke functions of the parent automaton in child automaton functions ? Or in this case we are needed to use sub's too ?

For example:

```
fun remove(): void
{
    //some code
    var index = self.parent.indexof(lastReturned);
    //some code
}
```
indexOf - this is fun.

dog-m replied:
Yes, only calls to `sub`routines are allowed in any of `fun`ctions. Even when calling parent methods.

dog-m commented on spec/java/util/LinkedList.lsl:
```diff
@@ -664,3 +669,58 @@ automaton ListItr: int(
+)
+automaton DescendingIterator: int(
+	//Do we can write in such way or not ??
+	var itr = new ListItr(state=Created,
```
Automata variables cannot have references to other automata.

alexeev509 replied:
@dog-m But how i must work in such situation ? I must dublicate hasPrevious(), previous() and remove() realization ? 

```
private class DescendingIterator implements Iterator<E> {
        private final ListItr itr = new ListItr(size());
        public boolean hasNext() {
            return itr.hasPrevious();
        }
        public E next() {
            return itr.previous();
        }
        public void remove() {
            itr.remove();
        }
    }
```

dog-m replied:
If it is unavoidable, then yes, code duplication is an acceptable solution while we cannot have automaton inheritance or templates.

alexeev509 committed:
LLSpliterator

Was added skeleton of the LLSpliterator automaton

alexeev509 committed:
count param to LIST_REMOVE action

Was extended functionality of the action LIST_REMOVE. Now it can remove continuous group of multiple items.

alexeev509 committed:
Update LinkedList.lsl

1. Was added new param to invocation of the action LIST_REMOVE.
2. Were added @Final annotation to final fields of the automaton.
3. Were removed returned types in signature of the constructors. 
4. Was added ListItr constructor in state transitions.

alexeev509 committed:
Subs place

Was changed place of the subs.

alexeev509 committed:
Object to any and invocations of funs to subs

1. Was changed "object" type to "any".
2. Were change invocation of the functions to invocations of the subs in the body of the fun's of LinkedList automaton.

alexeev509 committed:
curly braces

It was brought to similar style of the curly braces: Every curly brace on the new line.

alexeev509 committed:
Update LinkedList.lsl

1. Was added index param at this line: "self.parent.linkAny(, e);".
2. Was fixed Incorrect field access separator at this line: "self.parent.linkAny(self.parent-size - 1, e);".
3. All increments and decrements were replaced in the such way: "nextIndex = nextIndex - 1", "expectedModCount = expectedModCount + 1".
4. Were removed annotations @Private before subs.

alexeev509 committed:
Spaces fixed

1. spacing: 4 Space symbols per 1 Tab symbol was added.
2.  Were added spacing around operators.

alexeev509 committed:
ListItr

Was change ListItr realization. Were added flags nextWasCalled and prevWasCalled for automaton. Was removed field "next" with object type from automaton.

alexeev509 committed:
Spaces

Were added 4 spaces for every line

alexeev509 committed:
Spaces replace

Replacing of the all spaces

alexeev509 committed:
Little fixes

Little fixes.
The main change - this is renaming of the sub. It had name "isElementIndex", now this name is "isValidIndex".

alexeev509 committed:
Little fixes.

1.  ArrayObject was replaced by array<any>.
2. "parent = self" was removed in the creating of the object.
3. LinkedList clone fun was fixed.
4. Was added "action" word in the "previous()" method.
5. Were changte methods of the "DescendingIterator" automaton.
